calico/openstack/mech_calico.py

```diff
@@ -468,33 +497,41 @@ def periodic_resync_thread(self):
         it with etcd, ensuring that the etcd database and Neutron are in
         synchronization with each other.
         """
-        # The very first thing we do is sleep for our startup interval before
-        # initializing our state.
-        eventlet.sleep(STARTUP_DELAY_SECS)
-        self._init_state()
-
-        while True:
-            LOG.info("Attempting periodic resync.")
-
-            # Only do the resync logic if we're actually the master node.
-            if self.transport.is_master:
-                LOG.info("I am master: proceeding with resync")
-                context = ctx.get_admin_context()
-
-                try:
-                    # First, resync endpoints.
-                    self.resync_endpoints(context)
-
-                    # Second, profiles.
-                    self.resync_profiles(context)
-
-                    # Now, set the config flags.
-                    self.transport.provide_felix_config()
-                except Exception:
-                    LOG.exception("Error in periodic resync thread.")
-
-            # Reschedule ourselves.
-            eventlet.sleep(RESYNC_INTERVAL_SECS)
+        try:
+            LOG.info("Periodic resync thread started")
+            while self._epoch == expected_epoch:
+                # Only do the resync logic if we're actually the master node.
+                if self.transport.is_master:
+                    LOG.info("I am master: doing periodic resync")
+                    context = ctx.get_admin_context()
+
+                    try:
+                        # First, resync endpoints.
+                        self.resync_endpoints(context)
+
+                        # Second, profiles.
+                        self.resync_profiles(context)
+
+                        # Now, set the config flags.
+                        self.transport.provide_felix_config()
+                    except Exception:
+                        LOG.exception("Error in periodic resync thread.")
+                    # Reschedule ourselves.
+                    eventlet.sleep(RESYNC_INTERVAL_SECS)
+                else:
+                    # Shorter sleep interval before we check if we've become
+                    # the master.  Avoids waiting a whole RESYNC_INTERVAL_SECS
+                    # if we just miss the master update.
+                    eventlet.sleep(MASTER_CHECK_INTERVAL_SECS)
+        except:
+            # TODO Should we tear down the process.
+            LOG.exception("Periodic resync thread died!")
+            if self.transport:
+                # Stop the transport so that we give up the mastership.
+                self.transport.stop()
+            raise
```

Comment:
Makes sure we try to log any BaseExceptions even if we don't want to catch them.  I've found it useful to do this sort of log around any worker thread in the past.  I think eventlet logs such failures but they go to stdout/err so we don't see them?